File: 프로그래머스/2/49994. 방문 길이/방문 길이.py
# level 2
def check(dirs):
    start_x = 5
    start_y = 5
    passed = []
    new_road_length = 0
    for dir in dirs:
        before = [start_x, start_y]
        if dir == 'U':
            if start_y + 1 > 10:
                continue
            else:
                tuple = (before, [start_x, start_y+1])
                if tuple in passed or tuple[::-1] in passed:
                    start_y += 1
                    continue
                else:
                    start_y += 1
                    passed.append(tuple)
                    new_road_length += 1
        elif dir == 'D':
            if start_y - 1 < 0 :
                continue
            else:
                tuple = (before, [start_x, start_y-1])
                if tuple in passed or tuple[::-1] in passed:
                    start_y -= 1
                    continue
                else:
                    start_y -= 1
                    passed.append(tuple)
                    new_road_length += 1
        elif dir == 'L':
            if start_x - 1 < 0:
                continue
            else:
                tuple = (before, [start_x-1, start_y])
                if tuple in passed or tuple[::-1] in passed:
                    start_x -= 1
                    continue
                else:
                    start_x -= 1
                    passed.append(tuple)
                    new_road_length += 1
        elif dir == 'R':
            if start_x + 1 > 10:
                continue
            else:
                tuple = (before, [start_x+1, start_y])
                if tuple in passed or tuple[::-1] in passed:
                    start_x += 1
                    continue
                else:
                    start_x += 1
                    passed.append(tuple)
                    new_road_length += 1
    print(passed)
    return new_road_length

def solution(dirs):
    
    return check(dirs)

File: 프로그래머스/2/49994. 방문 길이/방문 길이.java
import java.util.*;
class Solution {
    public int solution(String dirs) {
        int answer = 0;

        Set<List<Integer>> visited = new HashSet<>();
        int start_x = 5;
        int start_y = 5;

        for(char dir : dirs.toCharArray()) {
            int end_x = start_x;
            int end_y = start_y;

            switch(dir) {
                case 'U':
                    if(start_y + 1 <=10) end_y++;
                    break;
                case 'D':
                    if(start_y -1 >= 0) end_y--;
                    break;
                case 'R':
                    if(start_x + 1 <= 10) end_x++;
                    break;
                case 'L':
                    if(start_x -1 >= 0) end_x--;
                    break;
            }

            if(start_x != end_x || start_y != end_y) {
                List<Integer> path = Arrays.asList(start_x,start_y,end_x,end_y);
                List<Integer> rev_path = Arrays.asList(end_x,end_y,start_x,start_y);

                if(visited.add(path) && visited.add(rev_path)) answer++;

                start_x = end_x;
                start_y = end_y;
            }
        }
        return answer;
    }
}
